File: js/zellular.ts
import * as redis from 'redis';
import * as url from 'url';
import { timeout } from './utils';

type ZellularBatchTx  = {
    type: "AgentRegister" | "CreateDepositAddress" | "Deposit" | "AddWithdraw" | "ApproveWithdraw" | "TransferWithdraw",
    data: any
}

export class Zellular {
	private appName: string;
	private baseUrl: string;
	private thresholdPercent: number;
	private r: redis.RedisClientType;

	constructor(appName: string, baseUrl: string, thresholdPercent: number = 67) {
		this.appName = appName;
		this.baseUrl = baseUrl;
		this.thresholdPercent = thresholdPercent;

		const parsedUrl = url.parse(baseUrl);
		this.r = redis.createClient({
			url: `redis://${parsedUrl.hostname}:${parsedUrl.port}`,
		});
		this.r.connect();
	}

	async *batches(after: number = 0): AsyncGenerator<[any, number], void, unknown> {
		if (after < 0) {
			throw new Error("after should be equal or bigger than 0");
		}
		while (true) {
			const batches = await this.r.lRange(this.appName, after, after + 100);
			for (let batch of batches) {
				after += 1;
				yield [batch, after];
			}
			await timeout(100); // Sleep for 0.1s
		}
	}

	async getLastFinalized() {
		const length = await this.r.lLen(this.appName);
		return { index: length };
	}

	async send(batch: ZellularBatchTx[], blocking: boolean = false) {
		let index;
		if (blocking) {
			index = await this.getLastFinalized();
		}

		await this.r.rPush(this.appName, JSON.stringify(batch));

		if (!blocking) {
			return;
		}

		let receivedBatch;
		for await ([receivedBatch, index] of this.batches(index)) {
			const parsedBatch = JSON.parse(receivedBatch);
			if (batch === parsedBatch) {
				return index;
			}
		}
	}
}

if (require.main === module) {
	(async () => {
		const verifier = new Zellular("custody_service_deposits", "http://localhost:6379");
		for await (const [batch, index] of verifier.batches()) {
			const txs = JSON.parse(batch);
			for (let i = 0; i < txs.length; i++) {
				console.log(index, i, txs[i]);
			}
		}
	})()
}

File: custody_service/zellular.py
import redis, json, asyncio, urllib, time

class Zellular:
    def __init__(self, app_name, base_url, threshold_percent=67):
        self.app_name = app_name
        self.base_url = base_url
        self.threshold_percent = threshold_percent
        url = urllib.parse.urlparse(base_url)
        self.r = redis.Redis(host=url.hostname, port=url.port, db=0)

    def batches(self, after=0):
        assert after >= 0, "after should be equal or bigger than 0"
        while True:
            batches = self.r.lrange(self.app_name, after, after + 100)
            for batch in batches:
                print(batch)
                after += 1
                yield batch, after
            time.sleep(0.1)

    def get_last_finalized(self):
        return { "index": self.r.llen(self.app_name) }

    def send(self, batch, blocking=False):
        if blocking:
            index = self.get_last_finalized()["index"]

        self.r.rpush(self.app_name, json.dumps(batch))

        if not blocking:
            return

        for received_batch, index in self.batches(after=index):
            received_batch = json.loads(received_batch)
            if batch == received_batch:
                return index


if __name__ == "__main__":
    verifier = Zellular("simple_app", "http://localhost:6379")
    for batch, index in verifier.batches():
        txs = json.loads(batch)
        for i, tx in enumerate(txs):
            print(index, i, tx)
